pytensor/link/jax/dispatch/scan.py

```diff
@@ -1,159 +1,209 @@
-            max_pos = max(pos_taps) if pos_taps else 0
-            init_slice = seq[: max_neg + max_pos]
-            mit_sot_in_slices.append(init_slice)
-
-        sit_sot_in_slices = [seq[0] for seq in scan_args.outer_in_sit_sot]
+    # Optimize inner graph
+    rewriter = op.mode_instance.optimizer
+    rewriter(op.fgraph)
+    scan_inner_func = jax_funcify(op.fgraph, **kwargs)
+
+    def scan(*outer_inputs):
+        # Extract JAX scan inputs
+        outer_inputs = list(outer_inputs)
+        n_steps = outer_inputs[0]  # JAX `length`
+        seqs = op.outer_seqs(outer_inputs)  # JAX `xs`
+
+        mit_sot_init = []
+        for tap, seq in zip(op.info.mit_sot_in_slices, op.outer_mitsot(outer_inputs)):
+            init_slice = seq[: abs(min(tap))]
+            mit_sot_init.append(init_slice)
+
+        sit_sot_init = [seq[0] for seq in op.outer_sitsot(outer_inputs)]
 
         init_carry = (
-            mit_mot_in_slices,
-            mit_sot_in_slices,
-            sit_sot_in_slices,
-            scan_args.outer_in_shared,
-            scan_args.outer_in_non_seqs,
-        )
+            mit_sot_init,
+            sit_sot_init,
+            op.outer_shared(outer_inputs),
+            op.outer_non_seqs(outer_inputs),
+        )  # JAX `init`
+
+        def jax_args_to_inner_func_args(op, carry, x):
+            """Convert JAX scan arguments into format expected by scan_inner_func.
+
+            scan(carry, x) -> scan_inner_func(seqs, mit_sot, sit_sot, shared, non_seqs)
+            """
 
-        def jax_args_to_inner_scan(op, carry, x):
-            # `carry` contains all inner-output taps, non_seqs, and shared
-            # terms
+            # `carry` contains all inner taps, non_seqs, and shared terms
             (
-                inner_in_mit_mot,
-                inner_in_mit_sot,
-                inner_in_sit_sot,
-                inner_in_shared,
-                inner_in_non_seqs,
+                inner_mit_sot,
+                inner_sit_sot,
+                inner_shared,
+                inner_non_seqs,
             ) = carry
 
-            # `x` contains the in_seqs
-            inner_in_seqs = x
+            # `x` contains the inner sequences
+            inner_seqs = x
 
-            # `inner_scan_inputs` is a list with the following composite form:
-            # inner_in_seqs
-            # + sum(inner_in_mit_mot, [])
-            # + sum(inner_in_mit_sot, [])
-            # + inner_in_sit_sot
-            # + inner_in_shared
-            # + inner_in_non_seqs
-            inner_in_mit_sot_flatten = []
-            for array, index in zip(inner_in_mit_sot, scan_args.mit_sot_in_slices):
-                inner_in_mit_sot_flatten.extend(array[jnp.array(index)])
+            mit_sot_flatten = []
+            for array, index in zip(inner_mit_sot, op.info.mit_sot_in_slices):
+                mit_sot_flatten.extend(array[jnp.array(index)])
 
+            # Concatenate lists
             inner_scan_inputs = sum(
```

Comment:
```suggestion
            inner_scan_inputs = [
                *inner_seqs,
                *mit_sot_flatten,
                *inner_sit_sot,
                *inner_shared,
                *inner_non_seqs,
            ]
```
I find this less scary and more immediate (GitHub doesn't let me apply the suggestion to the whole line of code so be careful if you apply it)

Reply:
Sure ;)